src/libraries/System.Text.RegularExpressions/gen/UpgradeToRegexGeneratorAnalyzer.cs

```diff
@@ -0,0 +1,290 @@
+    {
+        private const string RegexTypeName = "System.Text.RegularExpressions.Regex";
+        private const string RegexGeneratorTypeName = "System.Text.RegularExpressions.RegexGeneratorAttribute";
+        private const string RegexOptionsTypeName = "System.Text.RegularExpressions.RegexOptions";
+
+        internal const string PatternIndexName = "PatternIndex";
+        internal const string RegexOptionsIndexName = "RegexOptionsIndex";
+
+        /// <inheritdoc />
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DiagnosticDescriptors.UseRegexSourceGeneration);
+
+        /// <inheritdoc />
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+
+            context.RegisterCompilationStartAction(async compilationContext =>
+            {
+                Compilation compilation = compilationContext.Compilation;
+
+                if (!ProjectSupportsRegexSourceGenerator(compilation, out INamedTypeSymbol? regexTypeSymbol))
+                {
+                    return;
+                }
+
+                if (await ProjectUsesTopLevelStatements(compilation, compilationContext.CancellationToken).ConfigureAwait(false))
+                {
+                    return;
+                }
+
+                INamedTypeSymbol regexOptionsTypeSymbol = compilation.GetTypeByMetadataName(RegexOptionsTypeName);
+                if (regexOptionsTypeSymbol == null)
+                {
+                    return;
+                }
+
+                HashSet<IMethodSymbol> methodsWithFourParameters = GetMethodSymbolHash(compilation, regexTypeSymbol,
+                    new HashSet<string> { "Count", "EnumerateMatches", "IsMatch", "Match", "Matches", "Split" });
+
+                HashSet<IMethodSymbol> replaceStaticMethods = GetMethodSymbolHash(compilation, regexTypeSymbol, new HashSet<string> { "Replace" });
+
+                // Register analysis of calls to the Regex constructors
+                compilationContext.RegisterOperationAction(context => AnalyzeObjectCreation(context, regexTypeSymbol, regexOptionsTypeSymbol), OperationKind.ObjectCreation);
+
+                // Register analysis of calls to Regex static methods
+                compilationContext.RegisterOperationAction(context => AnalyzeInvocation(context, regexTypeSymbol, regexOptionsTypeSymbol, methodsWithFourParameters, replaceStaticMethods), OperationKind.Invocation);
+            });
+
+            static HashSet<IMethodSymbol> GetMethodSymbolHash(Compilation compilation, INamedTypeSymbol regexTypeSymbol, HashSet<string> methodNames)
+            {
+#pragma warning disable RS1024 // Compare symbols correctly
+                HashSet<IMethodSymbol> hash = new HashSet<IMethodSymbol>(SymbolEqualityComparer.Default);
```

Comment:
@mavasani I believe this is what you suggested to do in order to precalculate the methods I care about and store them in a hashset, but for some reason I keep getting build errors for RS1024 saying I'm not comparing symbols correctly even when passing the SymbolEqualityComparer. Any idea on how to correctly address that warning?

Reply:
Chatted with @mavasani offline, apparently this is a bug on the analyzer since this shouldn't warn. Bug is https://github.com/dotnet/roslyn-analyzers/issues/5804. I will add a comment to link this bug on the suppression so that we remember to remove it once we are ready to ingest the new Analyzers package in the repo.

Reply:
> Bug is https://github.com/dotnet/roslyn-analyzers/issues/5804. I will add a comment to link this bug on the suppression so that we remember to remove it once we are ready to ingest the new Analyzers package in the repo.

That bug was marked fixed in February.  Is that the right issue?  We've ingested new analyzer packages many times since then.

Reply:
@stephentoub Have you upgraded Microsoft.CodeAnalysis.Analyzers package as well? Do you reference it directly or get it via Microsoft.CodeAnalysis NuGet package?

Reply:
To my knowledge we don't reference it directly, just the NetAnalyzers package:
https://github.com/dotnet/runtime/blob/e96642078cfa9c68305ced1ee6accc1b0e0a9a9e/eng/Analyzers.props#L10
https://github.com/dotnet/runtime/blob/fdcedd873ddfa9f639f64481652d1f9c258e37b1/eng/Versions.props#L47

Reply:
RS1024 analyzers ships in Microsoft.CodeAnalysis.Analyzers NuGet package, which is a dependency of Microsoft.CodeAnalysis package. You either need to move to reference a newer version of Microsoft.CodeAnalysis package or add an explicit package reference to Microsoft.CodeAnalysis.Analyzers package to get the fix.

Reply:
We're referencing:
https://github.com/dotnet/runtime/blob/fdcedd873ddfa9f639f64481652d1f9c258e37b1/eng/Versions.props#L48
Is that not new enough?  Or maybe we're not actually using that version in all places that matter?

Reply:
So the version that flows in from Microsoft.CodeAnalysis package dependency is 3.3.3: https://github.com/dotnet/roslyn/blob/828ca46ab6ef59fc1747df27ff31b0fd86621669/eng/Versions.props#L77

https://www.nuget.org/packages/Microsoft.CodeAnalysis.Analyzers/ shows that version 3.3.3 was released about 7 months ago and won't have this fix. The only way for you to get the fix is to explicitly add a package reference to Microsoft.CodeAnalysis.Analyzers with the latest pre-release version ([3.3.4-beta1.22217.1](https://www.nuget.org/packages/Microsoft.CodeAnalysis.Analyzers/3.3.4-beta1.22217.1))

Reply:
Actually I checked and we are getting version 3.3.2 comming in as a transitive dependency from Microsoft.CodeAnalysis.CSharp.Workspaces which is a direct dependency of the project, but it is using a version which is repo-wide. In order for us not to have to upgrade that dependency repo-wide on this PR, and also in order to avoid going out of sync between roslyn assemblies, I have added a comment and a ToDo to remove the pragma once we update those repo-wide versions.